### PR TITLE
[ADT] Teach set_intersect to erase with iterators

### PR DESCRIPTION
**Pull request by kazutakahirata**
Without this patch, we erase an element in S1 by value even though we
have an interator pointing to it.  This patch tries to use erase(iter)
to avoid redundant lookups.

### CODE REVIEW

**Comment by nikic** [comment]
> Existing tests cover the new code path. Somewhat worrying is that the
existing "erase-by-value" code path is no longer covered. I am not
aware of any container that has erase(value) but not remove_if or
erase(iter).

I think my preference here would be to just always use the erase(iter) variant then. We can add back the erase(value) variant if we hit a use-case for it (and add test coverage for it then).

File: llvm/include/llvm/ADT/SetOperations.h [hunk 1]
```diff
@@ -60,6 +60,13 @@ template <class S1Ty, class S2Ty> void set_intersect(S1Ty &S1, const S2Ty &S2) {
   auto Pred = [&S2](const auto &E) { return !S2.count(E); };
   if constexpr (detail::HasMemberRemoveIf<S1Ty, decltype(Pred)>) {
     S1.remove_if(Pred);
+  } else if constexpr (detail::HasMemberEraseIter<S1Ty>) {
+    typename S1Ty::iterator Next;
+    for (typename S1Ty::iterator I = S1.begin(); I != S1.end(); I = Next) {
+      Next = std::next(I);
```
Comment (kuhar) [inline llvm/include/llvm/ADT/SetOperations.h:66]: nit: Just below we call `++I`. Do we need the `std::next` then?

**Comment by kazutakahirata** [comment]
> I think my preference here would be to just always use the erase(iter) variant then. We can add back the erase(value) variant if we hit a use-case for it (and add test coverage for it then).

OK.  The latest iteration uses erase(iter) without checking `detail::HasMemberEraseIter<S1Ty>`.

File: llvm/include/llvm/ADT/SetOperations.h (hunk 1 above)
Comment (kazutakahirata) [reply llvm/include/llvm/ADT/SetOperations.h:66]: We can use `std::next` there, too.  Are you suggesting replacing `++I` with `std::next` for consistency?

Comment (kuhar) [reply llvm/include/llvm/ADT/SetOperations.h:66]: I'd use `++I` for consistency if it's known to work in this context, otherwise use `next` in both places

Comment (kazutakahirata) [reply llvm/include/llvm/ADT/SetOperations.h:66]: Thank you for the clarification.  Here, I need to increment `I`, store the result in a another variable, and then call `S1.erase(I)` with the original value of `I`.  I'll use `std::next` in both places then.

Comment (kazutakahirata) [reply llvm/include/llvm/ADT/SetOperations.h:66]: Wait, @nikic suggested to drop the "erase-by-value" code path, so there is no "both places" anymore.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**